HV000151: A method overriding another method must not redefine the parameter constraint configuration

**I use** 
- reactivity (quarkus-resteasy-reactive-jackson, quarkus-rest-client-reactive-jackson)
- Java 11
- Quarkus 3.2.2.Final

When I try to run the application, , I end up with following error:

> Caused by: jakarta.validation.ConstraintDeclarationException: **HV000151: A method overriding another method must not redefine the parameter constraint configuration, but method TestServiceInterface$$CDIWrapper#doSomething(String) redefines the configuration of TestServiceInterface#doSomething(String)**.
	at org.hibernate.validator.internal.metadata.aggregated.rule.OverridingMethodMustNotAlterParameterConstraints.apply(OverridingMethodMustNotAlterParameterConstraints.java:24)
	at org.hibernate.validator.internal.metadata.aggregated.ExecutableMetaData$Builder.assertCorrectnessOfConfiguration(ExecutableMetaData.java:462)
	at org.hibernate.validator.internal.metadata.aggregated.ExecutableMetaData$Builder.build(ExecutableMetaData.java:380)
	at org.hibernate.validator.internal.metadata.aggregated.BeanMetaDataBuilder$BuilderDelegate.build(BeanMetaDataBuilder.java:260)
	at org.hibernate.validator.internal.metadata.aggregated.BeanMetaDataBuilder.build(BeanMetaDataBuilder.java:133)
	at org.hibernate.validator.internal.metadata.PredefinedScopeBeanMetaDataManager.createBeanMetaData(PredefinedScopeBeanMetaDataManager.java:155)
	at org.hibernate.validator.internal.metadata.PredefinedScopeBeanMetaDataManager.<init>(PredefinedScopeBeanMetaDataManager.java:100)
	at org.hibernate.validator.internal.engine.PredefinedScopeValidatorFactoryImpl.<init>(PredefinedScopeValidatorFactoryImpl.java:206)
	at org.hibernate.validator.PredefinedScopeHibernateValidator.buildValidatorFactory(PredefinedScopeHibernateValidator.java:42)
	at org.hibernate.validator.internal.engine.AbstractConfigurationImpl.buildValidatorFactory(AbstractConfigurationImpl.java:435)
	at io.quarkus.hibernate.validator.runtime.HibernateValidatorRecorder$2.created(HibernateValidatorRecorder.java:178)
	at io.quarkus.arc.runtime.ArcRecorder.initBeanContainer(ArcRecorder.java:79)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources844392269.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.ArcProcessor$generateResources844392269.deploy(Unknown Source)
	... 15 more


I have an interface, with parameter annotated `@ConvertGroup(to = Error.class)` and I have an implementation of that interface:

![image](https://github.com/quarkusio/quarkus/assets/18206019/55e62824-3fa4-4ce3-98d4-6dcdc6fe2ab3)
![image](https://github.com/quarkusio/quarkus/assets/18206019/6e014615-692b-4079-9a24-8570ba222aed)

I understand that there are $$CDIWrapper classes which are generated from the interface.
I have noticed, that when applying OverridingMethodMustNotAlterParameterConstraints, there is an issue that callable 
`TestServiceInterface$$CDIWrapper#doSomething(String)` has empty **groupConversions** while `TestServiceInterface#doSomething(String)` have exactly one... 

I think this is a bug. 
What is interesting, I tried to dump all generated classes with (`-Dquarkus.debug.generated-classes-dir=dump-classes`) and I can see, that the ConvertGroup annotation is present on that $$CDIWrapper class:

![image](https://github.com/quarkusio/quarkus/assets/18206019/cb112264-34da-40bb-84e0-be4f9316615f)

**Here is a project to reproduce the error https://github.com/dntichy/convert-group-minimal-example**




### Expected behavior

Exception should not be thrown and app should start properly

### Actual behavior

Exception is thrown upon startup

### How to Reproduce?

**Here is a project to reproduce the error https://github.com/dntichy/convert-group-minimal-example**


### Output of `java -version`

openjdk version "11.0.19" 2023-04-18 LTS
OpenJDK Runtime Environment Corretto-11.0.19.7.1 (build 11.0.19+7-LTS)
OpenJDK 64-Bit Server VM Corretto-11.0.19.7.1 (build 11.0.19+7-LTS, mixed mode)

### Quarkus version or git rev

3.2.2.Final

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -64,7 +64,7 @@
         <jakarta.annotation-api.version>3.0.0</jakarta.annotation-api.version>
         <jakarta.authentication-api>3.0.0</jakarta.authentication-api>
         <jakarta.authorization-api.version>2.1.0</jakarta.authorization-api.version>
-        <jakarta.el-api.version>5.0.1</jakarta.el-api.version>
+        <jakarta.el-api.version>6.0.1</jakarta.el-api.version>
         <jakarta.enterprise.cdi-api.version>4.1.0</jakarta.enterprise.cdi-api.version>
         <jakarta.inject-api.version>2.0.1</jakarta.inject-api.version>
         <jakarta.interceptor-api.version>2.2.0</jakarta.interceptor-api.version>
@@ -76,7 +76,7 @@
         <jakarta.resource-api.version>2.1.0</jakarta.resource-api.version>
         <jakarta.servlet-api.version>6.0.0</jakarta.servlet-api.version>
         <jakarta.transaction-api.version>2.0.1</jakarta.transaction-api.version>
-        <jakarta.validation-api.version>3.0.2</jakarta.validation-api.version>
+        <jakarta.validation-api.version>3.1.0</jakarta.validation-api.version>
         <jakarta.websocket-api.version>2.1.1</jakarta.websocket-api.version>
         <jakarta.ws.rs-api.version>3.1.0</jakarta.ws.rs-api.version>
         <jakarta.xml.bind-api.version>4.0.2</jakarta.xml.bind-api.version>

--- a/docs/src/main/asciidoc/validation.adoc
+++ b/docs/src/main/asciidoc/validation.adoc
@@ -585,7 +585,8 @@ so that we could register the appropriate classes for reflection.
 
 To configure the `ValidatorFactory`, use the exposed configuration properties and the CDI integration.
 
-Consequently, the only way to define constraints in Quarkus is by annotating your classes.
+While configuration of the validator factory is not possible through XML, constraint declaration is.
+Quarkus accepts constraints defined by annotations in your classes or in the `validation.xml`.
 
 === ValidatorFactory and native executables
 

--- a/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/ArcConstraintValidatorFactoryImpl.java
+++ b/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/ArcConstraintValidatorFactoryImpl.java
@@ -7,7 +7,7 @@ import jakarta.enterprise.context.Dependent;
 import jakarta.validation.ConstraintValidator;
 import jakarta.validation.ConstraintValidatorFactory;
 
-import org.hibernate.validator.internal.util.privilegedactions.NewInstance;
+import org.hibernate.validator.internal.util.actions.NewInstance;
 
 import io.quarkus.arc.Arc;
 import io.quarkus.arc.InstanceHandle;
@@ -32,7 +32,7 @@ public class ArcConstraintValidatorFactoryImpl implements ConstraintValidatorFac
             }
             return instance;
         }
-        return NewInstance.action(key, "ConstraintValidator").run();
+        return NewInstance.action(key, "ConstraintValidator");
     }
 
     @Override

--- a/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateBeanValidationConfigValidator.java
+++ b/extensions/hibernate-validator/runtime/src/main/java/io/quarkus/hibernate/validator/runtime/HibernateBeanValidationConfigValidator.java
@@ -8,7 +8,7 @@ import jakarta.validation.ValidatorFactory;
 
 import org.hibernate.validator.PredefinedScopeHibernateValidator;
 import org.hibernate.validator.PredefinedScopeHibernateValidatorConfiguration;
-import org.hibernate.validator.internal.engine.constraintvalidation.ConstraintValidatorFactoryImpl;
+import org.hibernate.validator.constraintvalidation.spi.DefaultConstraintValidatorFactory;
 
 import io.smallrye.config.validator.BeanValidationConfigValidator;
 
@@ -24,7 +24,7 @@ public class HibernateBeanValidationConfigValidator implements BeanValidationCon
                 .ignoreXmlConfiguration()
                 .builtinConstraints(constraints)
                 .initializeBeanMetaData(classesToBeValidated)
-                .constraintValidatorFactory(new ConstraintValidatorFactoryImpl())
+                .constraintValidatorFactory(new DefaultConstraintValidatorFactory())
                 .traversableResolver(new TraverseAllTraversableResolver());
 
         ConfigValidatorHolder.initialize(configuration.buildValidatorFactory());

--- a/integration-tests/hibernate-validator-resteasy-reactive/pom.xml
+++ b/integration-tests/hibernate-validator-resteasy-reactive/pom.xml
@@ -24,6 +24,14 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-rest-client</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-rest-client-jsonb</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
             <artifactId>quarkus-hibernate-validator</artifactId>
         </dependency>
         <dependency>
@@ -136,6 +144,32 @@
         <dependency>
             <groupId>io.quarkus</groupId>
             <artifactId>quarkus-rest-jaxb-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-rest-client-deployment</artifactId>
+            <version>${project.version}</version>
+            <type>pom</type>
+            <scope>test</scope>
+            <exclusions>
+                <exclusion>
+                    <groupId>*</groupId>
+                    <artifactId>*</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-rest-client-jsonb-deployment</artifactId>
             <version>${project.version}</version>
             <type>pom</type>
             <scope>test</scope>

--- a/integration-tests/hibernate-validator-resteasy-reactive/src/main/java/io/quarkus/it/hibernate/validator/restclient/RestClientEntity.java
+++ b/integration-tests/hibernate-validator-resteasy-reactive/src/main/java/io/quarkus/it/hibernate/validator/restclient/RestClientEntity.java
@@ -1,0 +1,30 @@
+package io.quarkus.it.hibernate.validator.restclient;
+
+import jakarta.validation.constraints.Max;
+import jakarta.validation.constraints.NotNull;
+
+public class RestClientEntity {
+    @Max(value = 5, groups = ErrorGroup.class)
+    public int number;
+    @NotNull
+    public String string;
+
+    public RestClientEntity() {
+    }
+
+    public RestClientEntity(int number, String string) {
+        this.number = number;
+        this.string = string;
+    }
+
+    @Override
+    public String toString() {
+        return "RestClientEntity{" +
+                "number=" + number +
+                ", string='" + string + '\'' +
+                '}';
+    }
+
+    public interface ErrorGroup {
+    }
+}

--- a/integration-tests/hibernate-validator-resteasy-reactive/src/main/java/io/quarkus/it/hibernate/validator/restclient/RestClientInterface.java
+++ b/integration-tests/hibernate-validator-resteasy-reactive/src/main/java/io/quarkus/it/hibernate/validator/restclient/RestClientInterface.java
@@ -1,0 +1,22 @@
+package io.quarkus.it.hibernate.validator.restclient;
+
+import jakarta.validation.Valid;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.groups.ConvertGroup;
+import jakarta.ws.rs.Consumes;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
+
+@RegisterRestClient
+@Path("/rest-client")
+public interface RestClientInterface {
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    RestClientEntity doSomething(
+            @Valid @ConvertGroup(to = RestClientEntity.ErrorGroup.class) @NotNull RestClientEntity parameter);
+}

--- a/integration-tests/hibernate-validator-resteasy-reactive/src/main/java/io/quarkus/it/hibernate/validator/restclient/RestClientTestResource.java
+++ b/integration-tests/hibernate-validator-resteasy-reactive/src/main/java/io/quarkus/it/hibernate/validator/restclient/RestClientTestResource.java
@@ -1,0 +1,31 @@
+package io.quarkus.it.hibernate.validator.restclient;
+
+import jakarta.inject.Inject;
+import jakarta.validation.ConstraintViolationException;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+import org.eclipse.microprofile.rest.client.inject.RestClient;
+
+@Path("/hibernate-validator/test-rest-client")
+public class RestClientTestResource {
+
+    @Inject
+    @RestClient
+    RestClientInterface restClient;
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public RestClientEntity doSomething() {
+        RestClientEntity entity = new RestClientEntity(9, "aaa");
+        try {
+            entity = restClient.doSomething(entity);
+            throw new IllegalStateException(
+                    "This point should not be reached. Validation should fail on the rest client call.");
+        } catch (ConstraintViolationException e) {
+            return entity;
+        }
+    }
+}

--- a/integration-tests/hibernate-validator-resteasy-reactive/src/main/java/io/quarkus/it/hibernate/validator/restclient/server/ServerResource.java
+++ b/integration-tests/hibernate-validator-resteasy-reactive/src/main/java/io/quarkus/it/hibernate/validator/restclient/server/ServerResource.java
@@ -1,0 +1,27 @@
+package io.quarkus.it.hibernate.validator.restclient.server;
+
+import jakarta.ws.rs.Consumes;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.core.MediaType;
+
+import io.quarkus.it.hibernate.validator.restclient.RestClientEntity;
+
+@Path("/rest-client")
+public class ServerResource {
+
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public String doSomething(String parameter) {
+        return parameter;
+    }
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public RestClientEntity doSomething() {
+        return new RestClientEntity(1, "aa");
+    }
+}

--- a/integration-tests/hibernate-validator-resteasy-reactive/src/main/resources/application.properties
+++ b/integration-tests/hibernate-validator-resteasy-reactive/src/main/resources/application.properties
@@ -1,2 +1,3 @@
 quarkus.locales=en,de
 quarkus.default-locale=en
+quarkus.rest-client."io.quarkus.it.hibernate.validator.restclient.RestClientInterface".url=${test.url}

--- a/integration-tests/hibernate-validator-resteasy-reactive/src/test/java/io/quarkus/it/hibernate/validator/RestClientTest.java
+++ b/integration-tests/hibernate-validator-resteasy-reactive/src/test/java/io/quarkus/it/hibernate/validator/RestClientTest.java
@@ -1,0 +1,18 @@
+package io.quarkus.it.hibernate.validator;
+
+import static io.restassured.RestAssured.given;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+public class RestClientTest {
+
+    @Test
+    public void fetchDefault() {
+        given().get("hibernate-validator/test-rest-client")
+                .then()
+                .statusCode(200);
+    }
+}

--- a/integration-tests/hibernate-validator/src/main/java/io/quarkus/it/hibernate/validator/HibernateValidatorTestResource.java
+++ b/integration-tests/hibernate-validator/src/main/java/io/quarkus/it/hibernate/validator/HibernateValidatorTestResource.java
@@ -48,6 +48,7 @@ import io.quarkus.it.hibernate.validator.injection.InjectedConstraintValidatorCo
 import io.quarkus.it.hibernate.validator.injection.InjectedRuntimeConstraintValidatorConstraint;
 import io.quarkus.it.hibernate.validator.injection.MyService;
 import io.quarkus.it.hibernate.validator.orm.TestEntity;
+import io.quarkus.it.hibernate.validator.xml.ValidationServiceBasedOnXmlConstraints;
 import io.quarkus.runtime.StartupEvent;
 
 @Path("/hibernate-validator/test")
@@ -63,6 +64,9 @@ public class HibernateValidatorTestResource
 
     @Inject
     EnhancedGreetingService enhancedGreetingService;
+
+    @Inject
+    ValidationServiceBasedOnXmlConstraints validationServiceBasedOnXmlConstraints;
 
     @Inject
     ZipCodeService zipCodeResource;
@@ -337,6 +341,17 @@ public class HibernateValidatorTestResource
         ResultBuilder result = new ResultBuilder();
 
         result.append(formatViolations(validator.validate(new Task())));
+
+        return result.build();
+    }
+
+    @GET
+    @Path("/constraints-defined-in-xml")
+    @Produces(MediaType.TEXT_PLAIN)
+    public String testConstraintsDefinedInXml() {
+        ResultBuilder result = new ResultBuilder();
+
+        result.append(formatViolations(validationServiceBasedOnXmlConstraints.validateSomeMyXmlBean()));
 
         return result.build();
     }

--- a/integration-tests/hibernate-validator/src/main/java/io/quarkus/it/hibernate/validator/xml/MyXmlBean.java
+++ b/integration-tests/hibernate-validator/src/main/java/io/quarkus/it/hibernate/validator/xml/MyXmlBean.java
@@ -1,0 +1,6 @@
+package io.quarkus.it.hibernate.validator.xml;
+
+public class MyXmlBean {
+    int id = 0;
+    String name;
+}

--- a/integration-tests/hibernate-validator/src/main/java/io/quarkus/it/hibernate/validator/xml/ValidationServiceBasedOnXmlConstraints.java
+++ b/integration-tests/hibernate-validator/src/main/java/io/quarkus/it/hibernate/validator/xml/ValidationServiceBasedOnXmlConstraints.java
@@ -1,0 +1,19 @@
+package io.quarkus.it.hibernate.validator.xml;
+
+import java.util.Set;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import jakarta.validation.ConstraintViolation;
+import jakarta.validation.Validator;
+
+@ApplicationScoped
+public class ValidationServiceBasedOnXmlConstraints {
+
+    @Inject
+    Validator validator;
+
+    public Set<ConstraintViolation<MyXmlBean>> validateSomeMyXmlBean() {
+        return validator.validate(new MyXmlBean());
+    }
+}

--- a/integration-tests/hibernate-validator/src/main/resources/META-INF/validation.xml
+++ b/integration-tests/hibernate-validator/src/main/resources/META-INF/validation.xml
@@ -1,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<validation-config
+        xmlns="https://jakarta.ee/xml/ns/validation/configuration"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="https://jakarta.ee/xml/ns/validation/configuration https://jakarta.ee/xml/ns/validation/validation-configuration-3.0.xsd"
+        version="3.0">
+    <constraint-mapping>META-INF/validation/constraints-my-xml-bean.xml</constraint-mapping>
+</validation-config>

--- a/integration-tests/hibernate-validator/src/main/resources/META-INF/validation/constraints-my-xml-bean.xml
+++ b/integration-tests/hibernate-validator/src/main/resources/META-INF/validation/constraints-my-xml-bean.xml
@@ -1,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<constraint-mappings
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="https://jakarta.ee/xml/ns/validation/mapping https://jakarta.ee/xml/ns/validation/validation-mapping-3.0.xsd"
+        xmlns="https://jakarta.ee/xml/ns/validation/mapping" version="3.0">
+
+    <bean class="io.quarkus.it.hibernate.validator.xml.MyXmlBean">
+        <field name="id">
+            <constraint annotation="jakarta.validation.constraints.Positive"/>
+        </field>
+        <field name="name">
+            <constraint annotation="jakarta.validation.constraints.NotNull"/>
+        </field>
+    </bean>
+</constraint-mappings>

--- a/integration-tests/hibernate-validator/src/test/java/io/quarkus/it/hibernate/validator/HibernateValidatorFunctionalityTest.java
+++ b/integration-tests/hibernate-validator/src/test/java/io/quarkus/it/hibernate/validator/HibernateValidatorFunctionalityTest.java
@@ -542,4 +542,16 @@ public class HibernateValidatorFunctionalityTest {
                 .then()
                 .body(is("passed"));
     }
+
+    @Test
+    void testConstraintsDefinedInXml() {
+        RestAssured.given()
+                .when()
+                .get("/hibernate-validator/test/constraints-defined-in-xml")
+                .then()
+                .statusCode(200)
+                .body(containsString("failed"),
+                        containsString("id (must be greater than 0)"),
+                        containsString("name (must not be null)"));
+    }
 }

--- a/pom.xml
+++ b/pom.xml
@@ -77,7 +77,7 @@
         <bytebuddy.version>1.15.11</bytebuddy.version> <!-- version controlled by Hibernate ORM's needs -->
         <hibernate-commons-annotations.version>7.0.3.Final</hibernate-commons-annotations.version> <!-- version controlled by Hibernate ORM's needs -->
         <hibernate-reactive.version>2.4.8.Final</hibernate-reactive.version> <!-- highly sensitive to Hibernate ORM upgrades -->
-        <hibernate-validator.version>8.0.2.Final</hibernate-validator.version>
+        <hibernate-validator.version>9.0.0.Final</hibernate-validator.version>
         <hibernate-search.version>7.2.3.Final</hibernate-search.version>
 
         <!-- Make sure to check compatibility between these 2 gRPC components before upgrade -->

